kratos/utilities/pointer_map_communicator.h

```diff
@@ -0,0 +1,623 @@
+ * gps will be stored, and applied after communication in their owning rank.
+ *
+ * Example:
+ * 1. Initial values
+ *      Node id   Rank    TEMPERATURE
+ *      1         0       10
+ *      2         1       20
+ *      3         2        5
+ *
+ * 2. Apply Update method with following Global Pointer map
+ *      Rank: 0
+ *      Map Key (using node id)         TEMPERATURE
+ *      1                               5
+ *      2                               10
+ *
+ *      Rank: 1
+ *      Map Key (using node id)         TEMPERATURE
+ *      1                               6
+ *      3                               2
+ *
+ *      Rank: 2
+ *      Empty Global Pointer map
+ *
+ * 3. Resulting values after SendAndApplyRemotely method call
+ *      Node id   Rank    TEMPERATURE
+ *      1         0       10 + 5 + 6 = 21
+ *      2         1       20 + 10 = 30
+ *      3         2       5 + 2 = 7
+ *
+ * Signature of TApplyFunctor:
+ *      void(TPointerDataType& rPointerDataTypeObject, const TValueDataType& NewValue)
+ *
+ *
+ * @tparam TPointerDataType
+ * @tparam TValueDataType
+ * @tparam TApplyFunctor
+ */
+template<class TPointerDataType, class TValueDataType, class TApplyFunctor>
+class ApplyProxy
+{
+public:
+    ///@name Type definitions
+    ///@{
+
+    using ProxyType = ApplyProxy<TPointerDataType, TValueDataType, TApplyFunctor>;
+
+    using TGPVector = GlobalPointersVector<TPointerDataType>;
+
+    using TGPDataMap = GlobalPointersUnorderedMap<TPointerDataType, TValueDataType>;
+
+    using TGPMapCommunicator = GlobalPointerMapCommunicator<TPointerDataType, TValueDataType>;
+
+    using GlobalPointerType = GlobalPointer<TPointerDataType>;
+
+    using GlobalPointerVectorType = std::vector<GlobalPointerType>;
+
+    using GlobalPointerMapType = std::unordered_map<int, GlobalPointerVectorType>;
+
+    using DataVectorType = std::vector<TValueDataType>;
+
+    using DataVectorMapType = std::unordered_map<int, DataVectorType>;
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    /**
+     * @brief Construct a new Apply Proxy object
+     *
+     * @param rApplyFunctor                         Thread safe update lambda method
+     * @param rPointerCommunicator                  Map communicator
+     */
+    ApplyProxy(
+        const TApplyFunctor& rApplyFunctor,
+        TGPMapCommunicator& rPointerCommunicator)
+        : mCurrentRank(rPointerCommunicator.GetMyPID()),
+          mrApplyFunctor(rApplyFunctor),
+          mrPointerCommunicator(rPointerCommunicator)
+    {
+        KRATOS_TRY
+
+        if (rPointerCommunicator.IsDistributed()) {
+
+            KRATOS_DEBUG_ERROR_IF(OpenMPUtils::IsInParallel() != 0)
+                << "Constructing a proxy in a parallel "
+                   "region is not allowed.\n";
+
+            // resizing for OMP parallel loops
+            const int number_of_threads = OpenMPUtils::GetNumThreads();
+            mNonLocalGlobalPointerMapsVector.resize(number_of_threads);
```

Comment:
same size as the number of threads??

Reply:
Yes for internal OMP-MPI hybrid parallelization. I think Kratos supports this, but trilinos or some other dependencies doesn't support it ryt? So I want to keep Kratos core supporting for hybrid parallelization..